src/DynamoCore/Graph/Nodes/PortModel.cs

```diff
@@ -440,6 +440,40 @@ internal string GetInputPortType()
             }
             return null;
         }
+
+        /// <summary>
+        /// Returns the string representation of the fully qualified typename
+        /// where possible for the port if it's an output port. This method currently
+        /// returns a valid type for only Zero Touch, Builtin and NodeModel nodes,
+        /// and returns null otherwise. The string representation of the type also
+        /// contains the rank information of the type, e.g. Point[], or var[]..[]. 
+        /// </summary>
+        /// <returns>output port type</returns>
+        internal string GetOutPortType()
+        {
+            if (PortType == PortType.Input) return null;
+
+            var ztNode = Owner as DSFunction;
+            if (ztNode != null)
+            {
+                var fd = ztNode.Controller.Definition;
+
+                string type = fd.ReturnType.ToString();
+              
+                return type;
+            }
+
+            var nmNode = Owner as NodeModel;
+            if (nmNode != null)
+            {
+                var classType = nmNode.GetType();
+
+                var outPortAttribute = classType.GetCustomAttributes().OfType<OutPortTypesAttribute>().FirstOrDefault();
+
+                return outPortAttribute?.PortTypes.ElementAt(Index);
```

Comment:
you can't guarantee that just because you have the `outputAtrribute` that your index won't be out of bounds. I think `ElementAt` might throw - you may want to bounds check first, or use a try catch.

Reply:
yeah, 3rd party dev or even us could forgot to match the number of `OutPortTypesAttribute`, I have fixed one instance in my other PR

Reply:
Added a try-catch block.